File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/mutation/SemanticMutationOperationGraph.java
package com.foundgine.core.semantic.mutation;

import java.util.*;

/** Canonical semantic mutation graph. */
public record SemanticMutationOperationGraph(List<SemanticMutationOperation> operations) {
    public SemanticMutationOperationGraph {
        if (operations == null || operations.isEmpty()) throw new IllegalArgumentException("A semantic mutation graph must contain at least one operation.");
        operations = List.copyOf(operations);
    }
    public List<SemanticMutationEffect> effects() { return operations.stream().flatMap(x -> x.effects().stream()).toList(); }
}

File: src/csharp/Foundgine.Core/Semantic/Mutation/SemanticMutationOperationGraph.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Mutation;

/// <summary>
/// Canonical semantic mutation graph. The graph describes what a mutation means
/// and how its effects depend on one another; physical execution is a later lowering.
/// </summary>
public sealed record SemanticMutationOperationGraph(
    IReadOnlyList<SemanticMutationOperation> Operations)
{
    public IEnumerable<SemanticMutationEffect> Effects =>
        Operations.SelectMany(x => x.Effects);
}